File: backend/Models/Trip.cs
﻿using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace intergalactica.Models;

public class Trip
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string Id { get; set; } = string.Empty;

    [BsonElement("source")]
    public string Source { get; set; } = string.Empty;
    [BsonElement("destination")]
    public string Destination { get; set; } = string.Empty;
    [BsonElement("source_dock")]
    public String SourceDock { get; set; } = string.Empty;
    [BsonElement("destination_dock")]
    public String DestinationDock { get; set; } = string.Empty;
    [BsonElement("shuttle")]
    public string Shuttle { get; set; } = string.Empty;
    [BsonElement("ticket_price")]
    public float TicketPrice { get; set; } = 0;

    //year is irrelevant, storing as a DateTime for ease of use
    //recurring trips will have the same date and time every year

    [BsonElement("departure")]
    public DateTime Departure { get; set; } = DateTime.MinValue;
    [BsonElement("arrival")]
    public DateTime Arrival { get; set; } = DateTime.MinValue;


}

File: frontend2/lib/models/trip.dart
class Trip {
  String? id;
  String? source;
  String? destination;
  String? sourceDock;
  String? destinationDock;
  DateTime? departure;
  DateTime? arrival;
  String? shuttle;
  double? ticketPrice;

  Trip({
    this.id,
    this.source,
    this.destination,
    this.sourceDock,
    this.destinationDock,
    this.departure,
    this.arrival,
    this.shuttle,
    this.ticketPrice,
  });

  static Trip fromJson(jsonDecode) {
    return Trip(
      id: jsonDecode['id'],
      source: jsonDecode['source'],
      destination: jsonDecode['destination'],
      sourceDock: jsonDecode['sourceDock'],
      destinationDock: jsonDecode['destinationDock'],
      departure: DateTime.parse(jsonDecode['departure']),
      arrival: DateTime.parse(jsonDecode['arrival']),
      shuttle: jsonDecode['shuttle'],
      ticketPrice: jsonDecode['ticketPrice'],
    );
  }

  static Map<String, dynamic> toJson(Trip trip) {
    return {
      'id': trip.id,
      'source': trip.source,
      'destination': trip.destination,
      'sourceDock': trip.sourceDock,
      'destinationDock': trip.destinationDock,
      'departure': trip.departure?.toIso8601String(),
      'arrival': trip.arrival?.toIso8601String(),
      'shuttle': trip.shuttle,
      'ticketPrice': trip.ticketPrice,
    };
  }

  @override
  String toString() {
    return 'Trip{id: $id, source: $source, destination: $destination, sourceDock: $sourceDock, destinationDock: $destinationDock, departure: $departure, arrival: $arrival, shuttle: $shuttle, ticketPrice: $ticketPrice}';
  }
}
